client/windows/CMessage.cpp

```diff
@@ -60,125 +60,227 @@ void CMessage::dispose()
+					 : trimmedLine.substr(wordEnd, *nextSplitPos - wordEnd);
+				nextWordWidth = fontPtr->getStringWidth(std::string(nextWord)) - findColorTags(nextWord);
+
+				if(lineWidth + nextWordWidth > maxLineWidth)
+				{
+					const auto resultLine = trimmedLine.substr(0, wordEnd);
+					// Trim splitSymbols from ending and place result line into result
+					result.back().line = std::make_pair(resultLine.data() - line.data(), resultLine.find_last_not_of(splitSymbol) + 1);
+					if(curColor.empty())
+					{
+						result.emplace_back();
+					}
+					else
+					{
+						result.back().closingTagNeeded = true;
+						result.emplace_back(std::optional<std::pair<size_t, size_t>>({ curColor.data() - line.data(), curColor.length() }));
+					}
+					splitBySymbol(trimmedLine.substr(wordEnd, trimmedLine.length() - wordEnd));
+					return;
+				}
+				else
 				{
-					opened = false;
-					color = "";
+					if(wordBegin == wordEnd)
+						lineWidth += fontPtr->getStringWidth(std::string(1, splitSymbol));
+					wordBegin = wordEnd + 1;
 				}
 			}
-			else
-				currPos -= symbolSize;
-		}
+			result.back().line = std::make_pair(trimmedLine.data() - line.data(), trimmedLine.length());
+		};
 
-		//non-blank line
-		if(currPos != 0)
-		{
-			ret.push_back(text.substr(0, currPos));
+	result.emplace_back();
+	splitBySymbol(line);
 
-			if(opened)
-				/* Close the brace for the current line. */
-				ret.back() += '}';
+	return result;
+}
 
-			text.erase(0, currPos);
-		}
-		else if(text[currPos] == 0x0a)
-		{
-			ret.push_back(""); //add empty string, no extra actions needed
-		}
+std::vector<std::string> CMessage::breakText(const std::string & text, size_t maxLineWidth, const EFonts font)
+{
+	std::vector<std::string> result;
+	if(maxLineWidth == 0)
+		return result;
+	if(!validateColorBraces(text))
+		return result;
+
+	// Detect end of line symbol. If '\r\n', remove '\r'.
+	// Texts with '\r\n' are very rare (Credits case), so rare string copying should not be a problem.
+	std::string preparedText;
+	auto preparedTextPtr = &text;
+	auto breakSymbol = '\n';
+	if(text.find("\r\n") != std::string::npos)
+	{
+		preparedText = text;
+		preparedTextPtr = &preparedText;
+		preparedText.erase(std::remove(preparedText.begin(), preparedText.end(), '\r'), preparedText.end());
+	}
+	else if(text.find('\r') != std::string::npos)
+	{
+		breakSymbol = '\r';
+	}
 
-		if(text.length() != 0 && text[0] == 0x0a)
-		{
-			/* Remove LF */
-			text.erase(0, 1);
-		}
-		else
-		{
-			// trim only if line does not starts with LF
-			// FIXME: necessary? All lines will be trimmed before returning anyway
-			boost::algorithm::trim_left_if(text, boost::algorithm::is_any_of(std::string(" ")));
-		}
+	// Split text by lines
+	auto possibleLines = getPossibleLines(*preparedTextPtr, 0, font, breakSymbol);
 
-		if(opened)
+	// Add additional empty lines if needed. In Case "....\n\n...."
```

Comment:
Perhaps this could be avoided by splitting input into sections to begin with? E.g. something like:
```cpp
std::vector<std::string> result;
std::vector<std::string> inputLines;
boost::split(inputLines, inputText, boost::is_any_of("\n"));
for (const auto & inputLine)
    result += getPossibleLines(inputLine)
```

With this approach there won't any need to handle line breaks in input text

Reply:
This approach may break the case where the colored string needs to be split "{..../n /n ....}"